metapool/abs_quant_sample_info.py

```
@@ -0,0 +1,334 @@
+	                <storage_liquid_lot_num>: {
+				        MASS_STORAGE_TUBE_ONLY_G_KEY: ##.##,
+				        DENSITY_STORAGE_LIQUID_G_ML_KEY: ##.##
+				    },
+		    MASS_STORAGE_TUBE_ONLY_G_KEY: ##.##
+		    },
+        <sample_density_g_ml_key>: ##.##
+        }
+
+    sample_density_g_ml_key: str
+        Dictionary key for the density of the sample type for the samples in
+        the input calc_df (usually stool).
+    storage_liquid_type: str
+        Type of storage liquid used with the samples in calc_df.
+
+    Returns
+    -------
+    calc_df: pandas.DataFrame
+        Extension of input calc_df with the following additional columns:
+            STORAGE_LIQUID_TYPE_KEY,
+            MASS_STORAGE_TUBE_ONLY_G_KEY
+            STORAGE_LIQUID_LOT_NUM_STR_KEY
+            MASS_STORAGE_TUBE_AND_STORAGE_LIQUID_BEFORE_SAMPLE_G_KEY
+            MASS_STORAGE_TUBE_AND_STORAGE_LIQUID_BEFORE_SAMPLE_MG_KEY
+            DENSITY_STORAGE_LIQUID_G_ML_KEY
+            DENSITY_STORAGE_LIQUID_G_ML_KEY
+            DENSITY_SAMPLE_G_ML_KEY
+            DENSITY_SAMPLE_KG_L_KEY
+    """
+
+    # yes, these steps could be chained together, but doing them separately
+    # makes it easier to see where errors come from if a key is missing
+    storage_liquids_dict = config_dict[STORAGE_LIQUID_TYPE_KEY]
+    storage_liquid_dict = storage_liquids_dict[storage_liquid_type]
+
+    calc_df[STORAGE_LIQUID_TYPE_KEY] = storage_liquid_type
+    calc_df[MASS_STORAGE_TUBE_ONLY_G_KEY] = \
+        storage_liquid_dict[MASS_STORAGE_TUBE_ONLY_G_KEY]
+
+    lot_nums_dict = storage_liquid_dict.get(STORAGE_LIQUID_LOT_NUM_STR_KEY)
+    calc_df[MASS_STORAGE_TUBE_AND_STORAGE_LIQUID_BEFORE_SAMPLE_G_KEY] = (
+        calc_df[STORAGE_LIQUID_LOT_NUM_STR_KEY].apply(
+            lambda x: lot_nums_dict.get(x).get(
+                MASS_STORAGE_TUBE_AND_STORAGE_LIQUID_BEFORE_SAMPLE_G_KEY)))
+    calc_df[MASS_STORAGE_TUBE_AND_STORAGE_LIQUID_BEFORE_SAMPLE_MG_KEY] = \
```

Do you need to vet the data type of the MASS_STORAGE_TUBE_AND_STORAGE_LIQUID_BEFORE_SAMPLE_G_KEY column before performing an arithmetic operation on them? I'm just wondering if that read_csv will accidentally open any of these columns as char objects instead of ints or floats and do we need to force data type above in the read_csv() call?

---

This column comes in via a file from the core, which we have agreed will follow a specific format.  If it does NOT follow that format, I actively do not want to allow the code to work with it--I want the code to fail loudly and messily so as to force an immediate conversation about why they broke the data transfer covenant :)